MuonAnalyzer/plugins/MuonFullAODAnalyzer.cc

```diff
@@ -336,17 +403,60 @@ MuonFullAODAnalyzer::analyze(const edm::Event& iEvent, const edm::EventSetup& iS
       FillTagBranches<reco::Muon> (tag.first,nt);
 
       std::vector<unsigned>::iterator it = std::find( trk_muon_map.first.begin(), trk_muon_map.first.end(), &probe - &tracks->at(0));
+      std::vector<unsigned>::iterator itdsa = std::find( trks_dsA_map.first.begin(), trks_dsA_map.first.end(), &probe - &tracks->at(0));
+      std::vector<unsigned>::iterator itdgl = std::find( trk_dglobal_map.first.begin(), trk_dglobal_map.first.end(), &probe - &tracks->at(0));
+      std::vector<unsigned>::iterator itcosmic = std::find( trk_cosmic_map.first.begin(), trk_cosmic_map.first.end(), &probe - &tracks->at(0));
       
-      if ( it == trk_muon_map.first.end() ){
+      if ( it == trk_muon_map.first.end()){
          reco::Muon fakeMuon;
          fakeMuon.setP4(mu2);
          FillProbeBranches<reco::Muon> (fakeMuon,nt,false);      
       }
+
       else{
          unsigned idx=std::distance(trk_muon_map.first.begin(),it);
          if (debug_>0) std::cout<<"successful probe pt "<<muons->at(trk_muon_map.second[idx]).pt()<<" eta "<<muons->at(trk_muon_map.second[idx]).eta()<<" phi "<<muons->at(trk_muon_map.second[idx]).phi()<<std::endl;
          FillProbeBranches<reco::Muon> (muons->at(trk_muon_map.second[idx]),nt,true);
       }
+
+      if ( itdsa == trks_dsA_map.first.end()){
+         reco::Muon fakedsaMuon;
+         fakedsaMuon.setP4(mu2);
+         FillProbeBranchesdSA<reco::Muon> (fakedsaMuon,nt,false);      
+      }
+
+      else{
+         unsigned itdsax=std::distance(trks_dsA_map.first.begin(),itdsa);
+         if (debug_>0) std::cout<<"successful probe dsA "<<muons->at(trks_dsA_map.second[itdsax]).pt()<<" eta "<<muons->at(trks_dsA_map.second[itdsax]).eta()<<" phi "<<muons->at(trks_dsA_map.second[itdsax]).phi()<<std::endl;
+         FillProbeBranchesdSA<reco::Muon> (muons->at(trks_dsA_map.second[itdsax]),nt,true);
```

Comment:
Hi @alfloren, I'm a bit unsure about this particular line of code. It seems to me like you have a mapping between generalTracks and dSA tracks in trks_dsA_map, but when you try to fill the probe branch in FillProbeBranchesdSA you call it with a reco::Muon object, and the one you choose is "muons->at(trks_dsA_map.second...". But the collection of muon objects does not necessarily match the collection of tracks, so I'm not sure what the track index will correspond to in the muon collection. Do you see what I mean?

Reply:
In fact there is another mapping trk_muon_map between generalTracks and reco::Muon objects which tells us what the relation between the two is. But in general a dSA track naturally does not have to have a reco::Muon object matched with it, so it seems like the index in muons->at(...) is a bit arbitrary.

Reply:
Actually the signature of the function is:

template<typename TRK>
inline void FillProbeBranchesdSA( const TRK & trk, NtupleContent & nt, bool passdSA)

so it expects a track, not a full muon object. We could change the signature of the function to accept a muon instead and use the two maps trks_dsA_map and trk_muon_map to get a muon object, but I'm not sure this is what we want to do (thinking at a higher-level here). In the case of dSA tracks, a lot of the time there won't be a muon object built for it, almost "by definition". So it's not really a fail condition if we can't find a match, right? I.e. it's not a "fake muon" just because we didn't find a match. I'm not sure if this makes sense, but maybe we can chat about it further.